### PR TITLE
builder submission performance improvements, cancel opt-in

### PR DESCRIPTION
**Pull request by metachris**
## 📝 Summary

Massive improvements to performance of block builder submissions. 

- Make cancellations opt-in (see #348), and for non-cancellation-enabled submissions, discard any bids with lower or equal value (skip validation)
- Do all local in-memory checks first, have Redis interaction as late as possible
- Grouped all Redis writes together and only write to Redis if necessary!
- Added a bunch of tests

Note!
- Cancellations need to be enabled with `ENABLE_BUILDER_CANCELLATIONS=1`
- Validation duration logging has changed from log field `duration` (used to be seconds) to `durationMs` (now in milliseconds)

Todo
- [ ] mark `wasSimulated=false` to discarded submissions db entries (see #355)
- [ ] add in-memory checks before going through redis (i.e. builder status, ...) - different PR

## ⛱ Motivation and Context

Block submission was slow, because our validation nodes are overloaded. 

Note the Flashbots relay receives about 500-700 submissions per slot.

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

### CODE REVIEW

File: services/api/service.go [hunk 1]
```diff
@@ -1335,9 +1333,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 
+	// Get the current (in-memory) top bid value
+	api.builderTopBidValueLock.RLock()
+	localTopBidValue := api.builderTopBidValue[payload.ParentHash()]
+	api.builderTopBidValueLock.RUnlock()
+
+	// Without cancellations, discard lower or similar value submissions to previous top bid
+	if !isCancellationEnabled {
+		// Check the local top bid value first
+		if localTopBidValue != nil && payload.Value().Cmp(localTopBidValue) < 1 {
```
Comment (Ruteri) [inline services/api/service.go:1344]: This check should take into account the parent & cl fields (randao in particular)

File: services/api/service.go [hunk 2]
```diff
@@ -1335,9 +1333,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
+		topBidValue, err := api.redis.GetTopBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+		if err != nil {
+			log.WithError(err).Error("failed to get top bid value from redis")
+		} else if topBidValue != nil {
+			log = log.WithField("topBidValue", topBidValue.String())
+
+			// Save locally for later in-memory lookup
+			if localTopBidValue == nil || localTopBidValue.Cmp(topBidValue) < 1 {
+				api.builderTopBidValueLock.Lock()
+				api.builderTopBidValue[payload.ParentHash()] = topBidValue
```
Comment (Ruteri) [inline services/api/service.go:1360]: `localTopBidValue` can hold an outdated payload, in which case this could overwrite the currently best payload

File: services/api/service.go [hunk 3]
```diff
@@ -1335,9 +1333,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		}
 	}()
 
+	// Get the current (in-memory) top bid value
+	api.builderTopBidValueLock.RLock()
+	localTopBidValue := api.builderTopBidValue[payload.ParentHash()]
```
Comment (Ruteri) [inline services/api/service.go:1338]: Is the parent hash the only thing that can differ for the payload to still be valid? Maybe a comment could be useful explaining why that's the case

File: services/api/service.go [hunk 4]
```diff
@@ -1453,24 +1489,36 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 
+	log = log.WithFields(logrus.Fields{
+		"timestampAfterUpdateTopBid": time.Now().UTC().UnixMilli(),
+		"newTopBidValue":             topBidValue,
+	})
+
+	// Update in-memory top bid value (if needed)
+	if topBidValue != nil || localTopBidValue.Cmp(topBidValue) < 1 {
+		api.builderTopBidValueLock.Lock()
+		api.builderTopBidValue[payload.ParentHash()] = topBidValue
```
Comment (Ruteri) [inline services/api/service.go:1510]: `api.builderTopBidValue[payload.ParentHash()]` could have changed in the meanwhile, and no longer be `localTopBidValue`

File: README.md [hunk 5]
```diff
@@ -155,7 +155,6 @@ redis-cli DEL boost-relay/sepolia:validators-registration boost-relay/sepolia:va
 * `DISABLE_PAYLOAD_DATABASE_STORAGE` - builder API - disable storing execution payloads in the database (i.e. when using memcached as data availability redundancy)
 * `DISABLE_LOWPRIO_BUILDERS` - reject block submissions by low-prio builders
 * `FORCE_GET_HEADER_204` - force 204 as getHeader response
-* `MEMCACHE_ALLOW_SAVING_FAIL` -- don't abort builder submission if memcache saving fails
```
Comment (Wazzymandias) [inline README.md:158]: Why was this removed? Is it because we were able to resolve the memcached connectivity issues on deploys?

File: services/api/service.go [hunk 6]
```diff
@@ -1201,24 +1199,36 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		return
 	}
 
+	// Timestamp check
+	expectedTimestamp := api.genesisInfo.Data.GenesisTime + (payload.Slot() * 12)
```
Comment (Wazzymandias) [inline services/api/service.go:1203]: nit: could we make a constant for `12` ? e.g. `SlotIntervalSeconds`

File: services/api/service.go [hunk 7]
```diff
@@ -1307,16 +1294,27 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 		log.WithError(err).Warn("could not verify builder signature")
 		api.RespondError(w, http.StatusBadRequest, "invalid signature")
 		return
 	}
 
+	// Reject new submissions once the payload for this slot was delivered - TODO: store in memory as well
+	slotLastPayloadDelivered, err := api.redis.GetStatsUint64(datastore.RedisStatsFieldSlotLastPayloadDelivered)
+	if err != nil && !errors.Is(err, redis.Nil) {
+		log.WithError(err).Error("failed to get delivered payload slot from redis")
+	} else if payload.Slot() <= slotLastPayloadDelivered {
```
Comment (Wazzymandias) [inline services/api/service.go:1312]: nit: the way the current conditional is structured, we would reach the `else if` clause when `err != nil` and `errors.Is(err, redis.Nil` 

That would cause us to perform a comparison with an invalid `slotLastPayloadDelivered` 

Since the value returned on error is 0, it probably has not caused issue, but it would be nice if we can make the logic clearer and avoid unnecessary comparison

File: services/api/service.go [hunk 8]
```diff
@@ -1335,9 +1333,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		}
 	}()
 
+	// Get the current (in-memory) top bid value
+	api.builderTopBidValueLock.RLock()
```
Comment (Wazzymandias) [inline services/api/service.go:1337]: I think it would be useful to encapsulate this in a function, e.g
```
func (api *RelayAPI) topBidValue(parentHash string) (value *big.Int, ok bool) {
	api.builderTopBidValueLock.RLock()
	defer api.builderTopBidValueLock.RUnlock()
	value, ok = api.builderTopBidValue[parentHash]
	return
}
```

File: services/api/service.go [hunk 9]
```diff
@@ -1335,9 +1333,47 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
+
+		// Then check the top bid value in redis
+		topBidValue, err := api.redis.GetTopBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+		if err != nil {
+			log.WithError(err).Error("failed to get top bid value from redis")
+		} else if topBidValue != nil {
+			log = log.WithField("topBidValue", topBidValue.String())
+
+			// Save locally for later in-memory lookup
+			if localTopBidValue == nil || localTopBidValue.Cmp(topBidValue) < 1 {
```
Comment (Wazzymandias) [inline services/api/service.go:1358]: Do we need to update the builderTopBidValue if it's the same value as `localTopBidValue` ? 

I'm wondering if we could do `localTopBidValue.Cmp(topBidValue) < 0` instead

File: datastore/redis.go [hunk 10]
```diff
@@ -407,17 +415,40 @@ func (r *RedisCache) UpdateTopBid(slot uint64, parentHash, proposerPubkey string
...
+	if err != nil {
+		return topBidValue, err
+	}
+
+	// Finally, update the top bid value
+	err = r.SetTopBidValue(slot, parentHash, proposerPubkey, topBidValue)
+	return topBidValue, err
+}
+
+func (r *RedisCache) SetTopBidValue(slot uint64, parentHash, proposerPubkey string, topBidValue *big.Int) (err error) {
```
Comment (Wazzymandias) [inline datastore/redis.go:440]: Could we add godoc for this function?

File: datastore/redis.go [hunk 11]
```diff
@@ -407,17 +415,40 @@ func (r *RedisCache) UpdateTopBid(slot uint64, parentHash, proposerPubkey string
...
+	keyTopBidValue := r.keyTopBidValue(slot, parentHash, proposerPubkey)
+	return r.client.Set(context.Background(), keyTopBidValue, topBidValue.String(), expiryBidCache).Err()
+}
+
+func (r *RedisCache) GetTopBidValue(slot uint64, parentHash, proposerPubkey string) (topBidValue *big.Int, err error) {
```
Comment (Wazzymandias) [inline datastore/redis.go:445]: Could we add godoc for this function?

File: services/api/service.go (hunk 7 above)
Comment (metachris) [reply services/api/service.go:1312]: great catch

File: services/api/service.go (hunk 8 above)
Comment (metachris) [reply services/api/service.go:1337]: on the fence about it. 3 lines and no repetition doesn't seem like it's worth it. if it's repeated 2 to 3 times then for sure.

File: services/api/service.go (hunk 4 above)
Comment (metachris) [reply services/api/service.go]: removed the comparison in the latest commit. after updating the bid in `UpdateTopBid`, the local value should _always_ be overwritten.

File: services/api/service.go (hunk 9 above)
Comment (metachris) [reply services/api/service.go]: removed the comparison, it should always overwrite, even if lower

File: services/api/service.go (hunk 2 above)
Comment (metachris) [reply services/api/service.go]: i've removed the local top bid value cache, as there's no way to reliably do it this way across multiple services. better use redis as source of truth also for this

File: datastore/redis.go [hunk 12]
```diff
@@ -386,38 +393,83 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
...
-			topBidBuilderPubkey = builderPubkey
+		if val.Cmp(prevTopBidValue) > 0 {
+			prevTopBidValue = val
+			prevTopBidBuilderPubkey = builderPubkey
 		}
 	}
 
-	if topBidBuilderPubkey == "" {
-		return ErrFailedUpdatingTopBidNoBids
+	// 3. Do nothing if there aren't wouldn't be any bids
```
Comment (Ruteri) [inline datastore/redis.go:419]: Do you mean "Do nothing if there aren't any bids"?

File: datastore/redis.go [hunk 13]
```diff
@@ -386,38 +393,83 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
...
+		return wasUpdated, err
 	}
+	wasUpdated = true
 
-	// Save the top bid
-	keyTopBid := r.keyCacheGetHeaderResponse(slot, parentHash, proposerPubkey)
-	return r.client.Set(context.Background(), keyTopBid, bidStr, expiryBidCache).Err()
+	// Finally, update the top bid value
+	keyTopBidValue := r.keyTopBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+	return wasUpdated, r.client.Set(context.Background(), keyTopBidValue, payload.Value().String(), expiryBidCache).Err()
```
Comment (Ruteri) [inline datastore/redis.go:459]: Could what's in `keyTopBidValue` change in the meanwhile, creating a race? Same for `keyTopBid`.
Shouldn't the update be a transaction too?

**Review by Ruteri — COMMENTED** [review]
It's become difficult to argue about correctness of `handleSubmitNewBlock` due to the size of the function. Maybe it's time to split it into smaller functions?

File: datastore/redis.go (hunk 13 above)
Comment (metachris) [reply datastore/redis.go]: yea should be in a transaction too

File: services/api/service.go [hunk 14]
```diff
@@ -1018,8 +1014,10 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 
 	// Check whether getPayload has already been called
 	slotLastPayloadDelivered, err := api.redis.GetStatsUint64(datastore.RedisStatsFieldSlotLastPayloadDelivered)
-	if err != nil && !errors.Is(err, redis.Nil) {
-		log.WithError(err).Error("failed to get delivered payload slot from redis")
+	if err != nil {
+		if !errors.Is(err, redis.Nil) {
+			log.WithError(err).Error("failed to get delivered payload slot from redis")
```
[the remark this answers is not in the document]
Comment (Wazzymandias) [reply services/api/service.go:1019]: previous discussion on this was here: https://github.com/flashbots/mev-boost-relay/pull/350#discussion_r1164093560

> then slotLastPayloadDelivered = 0 and the else if will fail
i don't think we should rely on values returned when an error occurs for the conditional logic here

File: services/api/service.go [hunk 15]
```diff
@@ -1415,65 +1433,45 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 	if err != nil {
 		log.WithError(err).Error("failed saving bidTrace in redis")
 		api.RespondError(w, http.StatusInternalServerError, err.Error())
 		return
 	}
 
-	// save execution payload (getPayload response)
-	err = api.redis.SaveExecutionPayload(payload.Slot(), payload.ProposerPubkey(), payload.BlockHash(), getPayloadResponse)
+	// 2. Save bid and recalculate top bid
+	wasUpdated, _, _, err := api.redis.SaveBidAndUpdateTopBid(payload, getPayloadResponse, getHeaderResponse, receivedAt, isCancellationEnabled)
```
[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go:1446]: First, for testing.

Also, i think we want to add these to the logs

File: datastore/redis.go [hunk 16]
```diff
@@ -386,38 +393,119 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
 	return r.client.Expire(context.Background(), keyLatestBidsValue, expiryBidCache).Err()
 }
 
-func (r *RedisCache) UpdateTopBid(slot uint64, parentHash, proposerPubkey string) (err error) {
-	// Get all builder's latest submission values
-	keyBidValues := r.keyBlockBuilderLatestBidsValue(slot, parentHash, proposerPubkey)
+func (r *RedisCache) SaveBidAndUpdateTopBid(payload *common.BuilderSubmitBlockRequest, getPayloadResponse *common.GetPayloadResponse, getHeaderResponse *common.GetHeaderResponse, reqReceivedAt time.Time, isCancellationEnabled bool) (wasTopBidUpdated bool, topBidValue *big.Int, topBidBuilderPubkey string, err error) {
```
[the remark this answers is not in the document]
Comment (metachris) [reply datastore/redis.go:396]: Makes sense to me

File: datastore/redis.go [hunk 17]
```diff
@@ -386,38 +393,119 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
...
+		// In cancellation mode, builder can overwrite his own last bid with lower value.
+		// First, compute newest top bid with latest values
+		for builderPubkey, bidValue := range bidValueIntMap {
+			if builderPubkey == payload.BuilderPubkey().String() {
+				bidValue = payload.Value() // overwrite previous value with new value
+			}
+			if bidValue.Cmp(topBidValue) > 0 {
+				topBidValue = bidValue
+				topBidBuilderPubkey = builderPubkey
+			}
```
[the remark this answers is not in the document]
Comment (Wazzymandias) [reply datastore/redis.go:465]: There was [another suggestion](https://github.com/flashbots/mev-boost-relay/pull/350#discussion_r1167120054) to create a separate struct to encapsulate the return values

I'm wondering if it would help if we did something like:
```
type TopBidResponse struct { // note: for illustrative purposes only
	updated             bool
	prevTopBid          *big.Int
	currentTopBid       *big.Int
	TopBidBuilderPubKey string
}

func (tbr *TopBidResponse) tryUpdateTopBid(input *big.Int) (*big.Int, bool) {
        const IsGreater = 1

	switch input.Cmp(tbr.currentTopBid) {
	case IsGreater:
		tbr.prevTopBid, tbr.currentTopBid = tbr.currentTopBid, input
		tbr.updated = true
	default:
		break
	}
	
	return tbr.prevTopBid, tbr.updated
}

func (tbr *TopBidResponse) TryUpdateTopBid(payload *common.BuilderSubmitBlockRequest, cancel bool, bidMap map[string]*big.Int) {
	payloadVal := payload.Value()
	if _, updated := tbr.tryUpdateTopBid(payloadVal); updated {
		return
	}

	if !cancel {
		return
	}

	for builderPubKey, bidValue := range bidMap {
		if builderPubKey == payload.BuilderPubkey().String() {
			bidValue.Set(payloadVal)
		}

		tbr.tryUpdateTopBid(bidValue)
	}
}

func (tbr *TopBidResponse) WasUpdated() bool {
	return tbr.updated
}
```

then within `SaveBidAndUpdateTopBid`, something like:
```
topBidResponse := &TopBidResponse{// TODO: Complete Me}
topBidResponse.TryUpdateTopBid(payload, isCancellationEnabled, bidValueIntMap)
if !topBidResponse.WasUpdated() {
  return topBidResponse, nil
} 
```

Since we'd be instantiating `TopBidResponse` based off the scope of the current request, we shouldn't need to read from Redis again if i'm understanding this correctly

[the remark this answers is not in the document]
Comment (Wazzymandias) [reply datastore/redis.go:465]: yeah the struct method was more as an example - i think it might be worth creating something like a `BidUpdater` interface (the name doesn't have to be BidUpdater) - imo the redis cache methods should hold cache specific logic, but atm it holds "bid specific" logic. I think it's worth separating those concerns so that it's easier for us to reason about and maintain

[the remark this answers is not in the document]
Comment (metachris) [reply datastore/redis.go:465]: I hear you. But I am wary of making too big changes, I'm more more looking for an incremental refactoring here right now. 

Also, the top bid specific logic is tied deeply into redis, which is the whole source of truth for this.

[the remark this answers is not in the document]
Comment (metachris) [reply datastore/redis.go:465]: Basically, updating a bd and computing the new top bid is just a bunch of redis operations

File: services/api/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go:1030]: @michaelneuder is right, we should keep the previous, proven logic, until there's a solid improvement, which the current attempt is clearly not (it's a mistake and i'm reverting ^^)

[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go:1030]: maybe the question here is whether `GetStatsUint64` should return an error if that stats thing wasn't found. i think it's the right thing to do tho

[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go]: reverted

[the remark this answers is not in the document]
Comment (metachris) [reply services/api/service.go]: thanks mike, great spot

File: datastore/redis.go (hunk 13 above)
Comment (metachris) [reply datastore/redis.go]: added a comment

File: datastore/redis.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (metachris) [reply datastore/redis.go]: done

File: common/types_spec.go [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+	}
+	gasLimit, err := json.Marshal(&struct {
+		RegisteredGasLimit uint64 `json:"registered_gas_limit,string"`
+	}{
+		RegisteredGasLimit: r.RegisteredGasLimit,
+	})
+	if err != nil {
+		return nil, err
+	}
+	gasLimit[0] = ','
```
[the remark this answers is not in the document]
Comment (metachris) [reply common/types_spec.go:254]: this PR just copied the code around, but yeah agree this feels weird. why do we even need a custom wrapper here?

File: datastore/redis.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (metachris) [reply datastore/redis.go]: okay, did a big refactoring taking all this into account! take a look https://github.com/flashbots/mev-boost-relay/blob/submitblock-performance-1/datastore/redis.go#L409

**Comment by metachris** [comment]
Deployed to Goerli, Sepolia and Zhejiang testnets

File: common/types_spec.go [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+func BuildGetHeaderResponse(payload *BuilderSubmitBlockRequest, sk *bls.SecretKey, pubkey *boostTypes.PublicKey, domain boostTypes.Domain) (*GetHeaderResponse, error) {
+	if payload == nil {
+		return nil, ErrMissingRequest
+	}
+
+	if sk == nil {
+		return nil, ErrMissingSecretKey
+	}
+
+	if payload.Bellatrix != nil {
```
Comment (michaelneuder) [inline common/types_spec.go:45]: dont need this anymore post capella, right?

File: common/types_spec.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+				Bellatrix: nil,
+			},
+			Bellatrix: nil,
+		}, nil
+	}
+	return nil, ErrEmptyPayload
+}
+
+func BuildGetPayloadResponse(payload *BuilderSubmitBlockRequest) (*GetPayloadResponse, error) {
+	if payload.Bellatrix != nil {
```
Comment (michaelneuder) [inline common/types_spec.go:77]: same comment as above.

File: common/types_spec.go [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+					AttesterSlashings:     capellaBlindedBlock.Message.Body.AttesterSlashings,
+					Attestations:          capellaBlindedBlock.Message.Body.Attestations,
+					Deposits:              capellaBlindedBlock.Message.Body.Deposits,
+					VoluntaryExits:        capellaBlindedBlock.Message.Body.VoluntaryExits,
+					SyncAggregate:         capellaBlindedBlock.Message.Body.SyncAggregate,
+					ExecutionPayload:      executionPayload.Capella.Capella,
+				},
+			},
+		}
+	} else if bellatrixBlindedBlock != nil {
```
Comment (michaelneuder) [inline common/types_spec.go:210]: same comment as above :-)

File: common/types_spec.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (michaelneuder) [reply common/types_spec.go:254]: we can leave for a follow up PR, but it seems like the reason gas limit is separate is because it is not a field in BuilderSubmitBlockRequest, so its not part of the unmarshall. but i haven't looked too closely.

File: datastore/redis.go [hunk 22]
```diff
@@ -386,38 +394,93 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
...
-	// Save the top bid
-	keyTopBid := r.keyCacheGetHeaderResponse(slot, parentHash, proposerPubkey)
-	return r.client.Set(context.Background(), keyTopBid, bidStr, expiryBidCache).Err()
+	state.WasTopBidUpdated = true
+	state.IsNewTopBid = payload.Value().Cmp(state.TopBidValue) == 0
+
+	// 6. Finally, update the global top bid value
+	keyTopBidValue := r.keyTopBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+	err = r.client.Set(context.Background(), keyTopBidValue, state.TopBidValue.String(), expiryBidCache).Err()
+	return state, err
```
Comment (michaelneuder) [inline datastore/redis.go:470]: beautiful!

**Review by michaelneuder — DISMISSED** [review]
lgtm. thanks for incorporating my comments :-)

File: common/types_spec.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (metachris) [reply common/types_spec.go:254]: let's definitely fix that in a follow-up PR

File: common/types_spec.go (hunk 19 above)
Comment (metachris) [reply common/types_spec.go:45]: yes, we can remove all the bellatrix stuff from the relay and mev-boost codebase soon!
